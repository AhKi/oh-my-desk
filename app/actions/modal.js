import { createActions } from 'redux-actions';
import * as TYPES from 'actions/constant/actionTypes';
import * as CATEGORY from 'actions/constant/actionCategory';

/**
 * Open and Close modal of each page.
 */
export const {
  modalOpen,
  modalClose,
} = createActions({
  [TYPES.MODAL_OPEN]: [
    /**
     * Open Modal in dispatched page if attach modal component.
     * @param:(ReactComponent | String) content: View Component of Modal or String of Modal Type.
     * @param:Object? props: injected props in Component
     * @returns {{content : *, props : *}}
     */
    (content, props) => ({ content, props }),
    (content, props, id) => {
      if (id) {
        return {
          category: CATEGORY.TARGET,
          target: [id],
          self: false,
        };
      }

      return { category: CATEGORY.SELF };
    },
  ],
  [TYPES.MODAL_CLOSE]: [
    () => {},
    () => ({ category: CATEGORY.SELF }),
  ],
});
