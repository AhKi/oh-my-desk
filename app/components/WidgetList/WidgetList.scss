.WidgetList {
  height: 100%;
}

.WidgetList__content {
  padding: 15px;
}
.WidgetList__list {
  margin: 20px;
}

.WidgetList__box {
  flex: 3 3 auto;
  margin: 20px;
}

.WidgetList__title {
  margin-bottom: 15px;
}

.WidgetList__add-btn {
  background-color: #e0a800;
  border: 1px solid rgba(0, 0, 0, .125);
  width: 100%;
  height: 45px;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  padding: .75rem 1.25rem;
  min-width: 350px;
}
