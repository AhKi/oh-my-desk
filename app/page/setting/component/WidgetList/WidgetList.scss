.WidgetList {
  height: 100%;
}

.WidgetList__content {
  padding: 15px;
}
.WidgetList__list {
  margin: 20px;
}

.WidgetList__box {
  flex: 3 3 auto;
  margin: 20px;
}

.WidgetList__title {
  margin-bottom: 15px;
}

.WidgetList__add-btn {
  background-color: #0000ff;
  border-radius: 5px;
  color: #ffffff;
  height: 45px;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  padding: .75rem 1.25rem;
}
