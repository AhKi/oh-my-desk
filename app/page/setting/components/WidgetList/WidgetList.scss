.WidgetList {
  height: 100%;
  background-color: $color-cool-gray20;
  color: $color-middle-gray-blue;
}

.WidgetList__content-header {
  color: $color-gray-blue;
  font-weight: 700;
  display: flex;
  align-items: center;
  justify-content: space-between;
}
.WidgetList__content {
  padding: 28px 40px;
}
.WidgetList__list {
  margin: 20px;
}

.WidgetList__box {
  flex: 3 3 auto;
  margin: 20px;
}

.WidgetList__title {
  margin-bottom: 15px;
}

.WidgetList__add-btn {
  border-radius: 5px;
  color: #ffffff;
  height: 45px;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  padding: .75rem 1.25rem;
}


.WidgetList__btn-box {
  padding: 25px 0;
  text-align: right;
}