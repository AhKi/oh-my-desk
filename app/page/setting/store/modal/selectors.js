import { createSelector } from 'reselect';

export const modalSelector = state => state.get('modal');
export const modalPropsSelector = createSelector(
  modalSelector,
  modal => modal.get('modalProps').toObject(),
);
export const modalTypeSelector = createSelector(
  modalSelector,
  modal => modal.get('modalType'),
);
