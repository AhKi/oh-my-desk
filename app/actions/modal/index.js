import { createActions } from 'redux-actions';
import * as TYPES from 'actions/actionTypes';
import * as CATEGORY from 'actions/category';

export const {
  modalOpen,
  modalClose,
} = createActions({
  [TYPES.MODAL_OPEN]: [
    (type, props) => ({ modalType: type, modalProps: props }),
    () => ({ category: CATEGORY.SELF }),
  ],
  [TYPES.MODAL_CLOSE]: [
    () => {},
    () => ({ category: CATEGORY.SELF }),
  ],
});
