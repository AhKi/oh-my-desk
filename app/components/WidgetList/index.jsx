import React from 'react';
import PropTypes from 'prop-types';
import * as IPC from 'constants/ipc';
import Header from 'components/Header';
import WidgetListBox from 'components/ListBox';
import * as MODAL from 'constants/modal';
import './WidgetList.scss';

const propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      favicon: PropTypes.string,
      id: PropTypes.string,
      isActive: PropTypes.bool,
      isIcon: PropTypes.bool,
      isOnTop: PropTypes.bool,
      name: PropTypes.string,
      position: PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number,
      }),
      size: PropTypes.shape({
        height: PropTypes.number,
        width: PropTypes.number,
      }),
      transparency: PropTypes.number,
      type: PropTypes.string,
      url: PropTypes.string,
    }),
  ),
  selectedId: PropTypes.string,
  onModalOpen: PropTypes.func,
  onStoreWidgetInfo: PropTypes.func,
  onSelectItem: PropTypes.func,
};

const defaultProps = {
  list: [],
  selectedId: '',
  onModalOpen() {},
  onStoreWidgetInfo() {},
  onSelectItem() {},
};

class WidgetList extends React.Component {
  constructor(props) {
    super(props);
    this.handleOpenModal = this.handleOpenModal.bind(this);
    this.handleSelectItem = this.handleSelectItem.bind(this);
  }

  componentDidMount() {
    window.ipcRenderer.send(IPC.WIDGET_INFO_REQUEST);
    window.ipcRenderer.on(IPC.WIDGET_INFO_RESULT,
      (response, result) => this.props.onStoreWidgetInfo(result));
  }

  handleOpenModal() {
    this.props.onModalOpen(MODAL.MAKE_WEB_WIDGET);
  }

  handleSelectItem(id) {
    window.ipcRenderer.send(IPC.WIDGET_INFO_REQUEST);
    this.props.onSelectItem(id);
  }

  render() {
    const {
      list,
      selectedId,
    } = this.props;

    return (
      <div className="WidgetList">
        <Header>
          <h1>Widget manage</h1>
        </Header>
        <div className="WidgetList__content">
          <div className="WidgetList__content-header">
            <h4>widget</h4>
          </div>
          <WidgetListBox
            list={list}
            selectedId={selectedId}
            onSelectItem={this.handleSelectItem}
          />
          <button
            className="WidgetList__add-btn"
            type="button"
            onClick={this.handleOpenModal}
          >
            <b><i className="fa fa-plus-square-o fa-lg" /> Add New Widget</b>
          </button>
        </div>
      </div>
    );
  }
}

WidgetList.propTypes = propTypes;
WidgetList.defaultProps = defaultProps;

export default WidgetList;
