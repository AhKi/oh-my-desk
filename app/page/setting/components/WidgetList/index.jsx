import React from 'react';
import PropTypes from 'prop-types';
import Header from 'setting/components/Header';
import Pagination from 'setting/components/Pagination';
import Select from 'setting/components/Select';
import * as CONST from 'constants/index';
import * as FILTER from 'constants/filter';
import * as IPC from 'constants/ipc';
import * as MODAL from 'constants/modal';
import WidgetListBox from './components/WidgetListBox';
import './WidgetList.scss';

const propTypes = {
  currentPage: PropTypes.number,
  filter: PropTypes.string,
  list: PropTypes.arrayOf(
    PropTypes.shape({
      favicon: PropTypes.string,
      id: PropTypes.string,
      isActive: PropTypes.bool,
      isIcon: PropTypes.bool,
      isOnTop: PropTypes.bool,
      name: PropTypes.string,
      size: PropTypes.shape({
        height: PropTypes.number,
        width: PropTypes.number,
      }),
      transparency: PropTypes.number,
      type: PropTypes.string,
      url: PropTypes.string,
    }),
  ),
  selectedId: PropTypes.string,
  totalNumber: PropTypes.number,
  onModalOpen: PropTypes.func,
  onStoreWidgetInfo: PropTypes.func,
  onSelectFilter: PropTypes.func,
  onSelectItem: PropTypes.func,
  onSelectPage: PropTypes.func,
  onUpdateInfoWithIPC: PropTypes.func,
};

const defaultProps = {
  currentPage: 1,
  filter: FILTER.LATEST,
  list: [],
  selectedId: '',
  totalNumber: 0,
  onModalOpen() {},
  onStoreWidgetInfo() {},
  onSelectFilter() {},
  onSelectItem() {},
  onSelectPage() {},
  onUpdateInfoWithIPC() {},
};

class WidgetList extends React.Component {
  constructor(props) {
    super(props);
    this.handleOpenModal = this.handleOpenModal.bind(this);
    this.handleSelectFilter = this.handleSelectFilter.bind(this);
    this.handleSelectItem = this.handleSelectItem.bind(this);
  }

  componentDidMount() {
    window.ipcRenderer.send(IPC.WIDGET_INFO_REQUEST);
    window.ipcRenderer.on(IPC.WIDGET_INFO_RESULT,
      (response, result) => this.props.onStoreWidgetInfo(result));
  }

  handleOpenModal() {
    this.props.onModalOpen(MODAL.MAKE_WEB_WIDGET);
  }

  handleSelectFilter(e) {
    this.props.onSelectFilter(e.target.value);
  }

  handleSelectItem(id) {
    window.ipcRenderer.send(IPC.WIDGET_INFO_REQUEST);
    this.props.onSelectItem(id);
  }

  render() {
    const {
      currentPage,
      filter,
      list,
      selectedId,
      totalNumber,
      onSelectPage,
      onUpdateInfoWithIPC,
      onModalOpen,
    } = this.props;
    const filterList = [FILTER.LATEST, FILTER.OLDEST, FILTER.ACTIVATED];
    const maxPage = Math.ceil(totalNumber / CONST.NUMBER_PER_PAGE);

    return (
      <div className="WidgetList">
        <Header>
          <h6>Widget manage</h6>
        </Header>
        <div className="WidgetList__content">
          <div className="WidgetList__content-header space-5x">
            <p>Activated widget</p>
            <Select
              items={filterList}
              value={filter}
              onChange={this.handleSelectFilter}
              className="Select"
            />
          </div>
          <WidgetListBox
            list={list}
            selectedId={selectedId}
            onSelectItem={this.handleSelectItem}
            onUpdateInfoWithIPC={onUpdateInfoWithIPC}
            onModalOpen={onModalOpen}
          />
          <div className="WidgetList__btn-box">
            <button
              className="Btn Btn--primary Btn--md"
              type="button"
              onClick={this.handleOpenModal}
            >
              <b>+ Add New Widget</b>
            </button>
          </div>
        </div>
        <Pagination
          currentPage={currentPage}
          maxPage={maxPage}
          onPageClick={onSelectPage}
        />
      </div>
    );
  }
}

WidgetList.propTypes = propTypes;
WidgetList.defaultProps = defaultProps;

export default WidgetList;
