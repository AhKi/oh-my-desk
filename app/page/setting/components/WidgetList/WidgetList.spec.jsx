import React from 'react';
import { shallow, mount } from 'enzyme';

import * as IPC from 'constants/ipc';
import * as MODAL from 'constants/modal';
import WidgetList from '.';

describe('<WidgetList />', () => {
  const ipcRenderer = {
    send: () => {},
    on: () => {},
  };
  window.ipcRenderer = ipcRenderer;

  it('should match to snapshot when render default', () => {
    const wrapper = shallow(<WidgetList />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call send and on action when componentDidMount', () => {
    const componentDidMount = jest.spyOn(WidgetList.prototype, 'componentDidMount');
    const send = jest.spyOn(ipcRenderer, 'send');
    const on = jest.spyOn(ipcRenderer, 'on');

    send.mockClear();
    on.mockClear();
    componentDidMount.mockClear();

    mount(<WidgetList />);

    expect(componentDidMount).toHaveBeenCalledTimes(1);
    expect(send).toHaveBeenCalledTimes(1);
    expect(send).toHaveBeenCalledWith(IPC.WIDGET_INFO_REQUEST);
    expect(on).toHaveBeenCalledTimes(1);
  });

  it('should call onModalOpen when call handleOnModal', () => {
    const onModalOpen = jest.fn();
    const wrapper = shallow(<WidgetList onModalOpen={onModalOpen} />);

    wrapper.instance().handleOpenModal();

    expect(onModalOpen).toHaveBeenCalledTimes(1);
    expect(onModalOpen).toHaveBeenCalledWith(MODAL.MAKE_WEB_WIDGET);
  });

  it('should call onSelectFilter when call handleSelectFilter', () => {
    const onSelectFilter = jest.fn();
    const wrapper = shallow(<WidgetList onSelectFilter={onSelectFilter} />);

    wrapper.instance().handleSelectFilter({
      target: {
        value: 'mock-value',
      },
    });

    expect(onSelectFilter).toHaveBeenCalledTimes(1);
    expect(onSelectFilter).toHaveBeenCalledWith('mock-value');
  });

  it('should call correct when call handleSelectItem', () => {
    const onSelectItem = jest.fn();
    const send = jest.spyOn(ipcRenderer, 'send');
    send.mockClear();
    const wrapper = shallow(<WidgetList onSelectItem={onSelectItem} />);

    wrapper.instance().handleSelectItem('mock-id');

    expect(send).toHaveBeenCalledTimes(2);
    expect(send).toHaveBeenCalledWith(IPC.WIDGET_INFO_REQUEST);
    expect(onSelectItem).toHaveBeenCalledTimes(1);
    expect(onSelectItem).toHaveBeenCalledWith('mock-id');
  });
});
