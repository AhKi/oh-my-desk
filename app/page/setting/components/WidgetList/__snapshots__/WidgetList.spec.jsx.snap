// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WidgetList /> should match to snapshot when render default 1`] = `
<div
  className="WidgetList"
>
  <Header>
    <h6>
      Widget manage
    </h6>
  </Header>
  <div
    className="WidgetList__content"
  >
    <div
      className="WidgetList__content-header space-5x"
    >
      <p>
        Activated widget
      </p>
      <Select
        className="Select"
        items={
          Array [
            "LATEST",
            "OLDEST",
            "ACTIVATED",
          ]
        }
        onChange={[Function]}
        value="LATEST"
      />
    </div>
    <WidgetListBox
      list={Array []}
      onModalOpen={[Function]}
      onSelectItem={[Function]}
      onUpdateInfoWithIPC={[Function]}
      selectedId=""
    />
    <div
      className="WidgetList__btn-box"
    >
      <button
        className="Btn Btn--primary Btn--md"
        onClick={[Function]}
        type="button"
      >
        <b>
          + Add New Widget
        </b>
      </button>
    </div>
  </div>
  <Pagination
    currentPage={1}
    maxPage={0}
    onPageClick={[Function]}
    unit={5}
  />
</div>
`;
